aten/src/ATen/native/cuda/GridSampler.cu

```diff
@@ -0,0 +1,782 @@
+	  return d >= 0 && d < D && h >= 0 && h < H && w >= 0 && w < W;
+	}
+
+  template<typename scalar_t>
+	static __forceinline__ __device__
+  void safe_add_2d(scalar_t *data, int h, int w,
+                   int sH, int sW, int H, int W,
+                   scalar_t delta) {
+    if (h >= 0 && h < H && w >= 0 && w < W) {
+      atomicAdd(data + h * sH + w * sW, delta);
+    }
+  }
+
+  template<typename scalar_t>
+	static __forceinline__ __device__
+  void safe_add_3d(scalar_t *data, int d, int h, int w,
+                   int sD, int sH, int sW, int D, int H, int W,
+                   scalar_t delta) {
+    if (d >= 0 && d < D && h >= 0 && h < H && w >= 0 && w < W) {
+      atomicAdd(data + d * sD + h * sH + w * sW, delta);
+    }
+  }
+
+	template <typename scalar_t>
+	__launch_bounds__(1024)
+	__global__ void grid_sampler_2d_kernel(
+	    const int nthreads,
+	    TensorInfo<scalar_t, int> input,
+	    TensorInfo<scalar_t, int> grid,
+	    TensorInfo<scalar_t, int> output,
+	    const GridSamplerPadding padding_mode) {
+
+	  int C = input.sizes[1];
+	  int inp_H = input.sizes[2];
+	  int inp_W = input.sizes[3];
+	  int out_H = grid.sizes[1];
+	  int out_W = grid.sizes[2];
+	  int inp_sN = input.strides[0];
+	  int inp_sC = input.strides[1];
+	  int inp_sH = input.strides[2];
+	  int inp_sW = input.strides[3];
+	  int grid_sN = grid.strides[0];
+	  int grid_sH = grid.strides[1];
+	  int grid_sW = grid.strides[2];
+	  int grid_sCoor = grid.strides[3];
+	  int out_sN = output.strides[0];
+	  int out_sC = output.strides[1];
+	  int out_sH = output.strides[2];
+	  int out_sW = output.strides[3];
+
+	  CUDA_KERNEL_LOOP(index, nthreads) {
+	  	const int w = index % out_W;
+	  	const int h = (index / out_W) % out_H;
+	  	const int n = index / (out_H * out_W);
+	  	const int grid_offset = n * grid_sN + h * grid_sH + w * grid_sW;
+
+	    // get the corresponding input x, y co-ordinates from grid
+	    scalar_t ix = grid.data[grid_offset];
+	    scalar_t iy = grid.data[grid_offset + grid_sCoor];
+
+	    // normalize ix, iy from [-1, 1] to [0, IH-1] & [0, IW-1]
+	    float ixf = ((ix + 1.f) / 2) * (inp_W - 1);
+	    float iyf = ((iy + 1.f) / 2) * (inp_H - 1);
+
+	    ix = static_cast<scalar_t>(ixf);
+	    iy = static_cast<scalar_t>(iyf);
+
+	    // get NE, NW, SE, SW pixel values from (x, y)
+	    int ix_nw = static_cast<int>(::floor(ixf));
+	    int iy_nw = static_cast<int>(::floor(iyf));
+	    int ix_ne = ix_nw + 1;
+	    int iy_ne = iy_nw;
+	    int ix_sw = ix_nw;
+	    int iy_sw = iy_nw + 1;
+	    int ix_se = ix_nw + 1;
+	    int iy_se = iy_nw + 1;
+
+	    // get surfaces to each neighbor:
+	    scalar_t nw = (ix_se - ix)    * (iy_se - iy);
+	    scalar_t ne = (ix    - ix_sw) * (iy_sw - iy);
+	    scalar_t sw = (ix_ne - ix)    * (iy    - iy_ne);
+	    scalar_t se = (ix    - ix_nw) * (iy    - iy_nw);
+
+	    // calculate bilinear weighted pixel value and set output pixel
+			if (padding_mode == GridSamplerPadding::Border) {
```

Comment:
Weirdly indented

Reply:
It's because you're mixing tabs and spaces in this file, it looks like.

Reply:
yeah. i fixed these in #10051 . I'll backport the fix here.